### PR TITLE
Add support for role icons

### PR DESCRIPTION
**Pull request by caneleex**
[contributing]: https://github.com/DV8FromTheWorld/JDA/wiki/5%29-Contributing

## Pull Request Etiquette

- [x] I have checked the PRs for upcoming features/bug fixes.
- [x] I have read the [contributing guidelines][contributing].

### Changes

- [ ] Internal code
- [x] Library interface (affecting end-user code) 
- [ ] Documentation
- [ ] Other: \_____ 

Closes Issue: NaN

## Description

this PR adds support for role icons. see https://github.com/discord/discord-api-docs/pull/3847

### CODE REVIEW

File: src/main/java/net/dv8tion/jda/api/managers/RoleManager.java [hunk 1]
```diff
@@ -56,6 +57,8 @@
     long HOIST       = 0x8;
     /** Used to reset the mentionable field */
     long MENTIONABLE = 0x10;
+    /** Used to reset the icon field */
+    long ICON        = 0x12;
```
Comment (DV8FromTheWorld) [inline src/main/java/net/dv8tion/jda/api/managers/RoleManager.java:61]:
```suggestion
    long ICON        = 0x20;
```
These are bit masks stored as hex.
We might want to change all of these to be bit masks built by `<<` to better represent that for future contributors.

**Review by DV8FromTheWorld — CHANGES_REQUESTED** [review]

**Comment by DV8FromTheWorld** [comment]
Looks pretty good to me. @MinnDevelopment?

**Comment by MinnDevelopment** [comment]
The docs PR is only a datamine PR for an experiment, so I would be against merging it at this time. See https://github.com/discord/discord-api-docs/pull/3815

**Comment by caneleex** [comment]
that's one of the reasons this PR is a draft. I haven't tested it

**Comment by DV8FromTheWorld** [comment]
Okay, going to mark as freezer until the feature is no longer experimental.
Additionally, this PR will need to be redirected at `v5` as we're closing down feature-additions on v4

**Comment by caneleex** [comment]
seems like the PR has been merged. I'll still wait until the feature is fully released to test it out properly

**Review by MinnDevelopment — CHANGES_REQUESTED** [review]

**Comment by caneleex** [comment]
it'd be great if someone with access to role icons (owning a server with boost tier 2+) could test this PR because I don't have one. (I think the PR should be fine though)

File: src/main/java/net/dv8tion/jda/api/managers/RoleManager.java [hunk 2]
```diff
@@ -273,6 +282,32 @@ default RoleManager setColor(@Nullable Color color)
...
+     *         or {@code null} to reset
+     *
+     * @return RoleManager for chaining convenience
+     */
+    @Nonnull
+    @CheckReturnValue
+    RoleManager setIcon(@Nullable Icon icon);
+
+    /**
+     * Sets the Unicode Emoji of this {@link net.dv8tion.jda.api.entities.Role Role}.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/managers/RoleManager.java:299]: The docs don't really make it clear to me what the difference between emoji and icon is

Comment (caneleex) [reply src/main/java/net/dv8tion/jda/api/managers/RoleManager.java:299]: I'd like to know that too without paying discord to see the feature - it's all I got from the docs: https://github.com/discord/discord-api-docs/blob/master/docs/resources/Guild.md#create-guild-role--post-guildsguildiddocs_resources_guildguild-objectroles

**Comment by fangming-ning-sp** [comment]
When will this be available? I'm waiting for it

File: src/main/java/net/dv8tion/jda/api/entities/Role.java [hunk 3]
```diff
@@ -291,6 +293,46 @@ default RoleAction createCopy()
...
+     * @return Possibly-null String containing the Role's icon hash-id.
+     *
+     * @since  4.3.1
+     */
+    @Nullable
+    String getIconId();
+
+    /**
+     * The URL of the {@link net.dv8tion.jda.api.entities.Role Role} icon image.
+     * If no icon has been set, this returns {@code null}.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/Role.java:310]:
```suggestion
     * If no icon has been set or an emoji is used in its place, this returns {@code null}.
```

File: src/main/java/net/dv8tion/jda/api/entities/Role.java [hunk 4]
```diff
@@ -291,6 +293,46 @@ default RoleAction createCopy()
     @Nonnull
     RoleTags getTags();
 
+    /**
+     * The Discord hash-id of the {@link net.dv8tion.jda.api.entities.Role Role} icon image.
+     * If no icon has been set, this returns {@code null}.
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/Role.java:298]:
```suggestion
     * If no icon has been set or an emoji is used in its place, this returns {@code null}.
```

**Review by MinnDevelopment — APPROVED** [review]
Only a small clarification in docs and then its good to go

File: src/main/java/net/dv8tion/jda/internal/handle/GuildRoleUpdateHandler.java [hunk 5]
```diff
@@ -121,6 +124,16 @@ protected Long handleInternally(DataObject content)
                             getJDA(), responseNumber,
                             role, wasMentionable));
         }
+
+        RoleIcon oldIcon = role.getIcon();
+        if (!Objects.equals(oldIcon.getIconId(), iconId) || !Objects.equals(oldIcon.getEmoji(), emoji))
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/internal/handle/GuildRoleUpdateHandler.java:129]: How about `RoleIcon#equals` and `RoleIcon#hashCode`?

Comment (caneleex) [reply src/main/java/net/dv8tion/jda/internal/handle/GuildRoleUpdateHandler.java:129]: I can create those, sure, but not sure how those would be helpful here as I don't have the new object before and creating a new one just for equality purposes is pointless

File: src/main/java/net/dv8tion/jda/api/entities/RoleIcon.java [hunk 6]
```diff
@@ -0,0 +1,92 @@
+package net.dv8tion.jda.api.entities;
```
Comment (MinnDevelopment) [inline src/main/java/net/dv8tion/jda/api/entities/RoleIcon.java:1]: This is missing copyright

**Review by MinnDevelopment — CHANGES_REQUESTED** [review]

**Review by MinnDevelopment — APPROVED** [review]

**Review by DV8FromTheWorld — APPROVED** [review]

**Pull request merged**